src/main/java/io/cryostat/discovery/AbstractNodeTypeAdapter.java

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright The Cryostat Authors
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or data
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.discovery;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import io.cryostat.platform.ServiceRef;
+import io.cryostat.platform.ServiceRef.AnnotationKey;
+import io.cryostat.platform.discovery.AbstractNode;
+import io.cryostat.platform.discovery.EnvironmentNode;
+import io.cryostat.platform.discovery.NodeType;
+import io.cryostat.platform.discovery.TargetNode;
+import io.cryostat.util.PluggableTypeAdapter;
+
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import dagger.Lazy;
+
+public class AbstractNodeTypeAdapter extends PluggableTypeAdapter<AbstractNode> {
```

Comment:
Ah, it's for this line basically:

https://github.com/andrewazores/cryostat/blob/555f38edbf06a71aae57c3f1915f02f13fc49590/src/main/java/io/cryostat/net/web/http/api/v2/DiscoveryPostHandler.java#L121

Without this type adapter, gson doesn't know how to deserialize the JSON into an `AbstractNode` - because it's Abstract, so it could be either an `EnvironmentNode` or a `TargetNode`, or possibly something else. Gson just doesn't have enough information about the types to figure this out automatically.

It probably will end up being used for serializing subtrees into JSON through gson to stick them into the database though, you're right on there.

Reply:
> It probably will end up being used for serializing subtrees into JSON through gson to stick them into the database though, you're right on there.

Well - Gson already does know how to do the serialization side for us. It does that in the existing `GET /api/v2.1/discovery` handler, for example. It's easier for Gson to serialize structures like this because it just uses reflection to read the names and types of fields, and then creates the JSON structure that way. So the fact that the `EnvironmentNode` or `TargetNode` it finds is an extension of an abstract class doesn't really matter.

There's still a benefit to having the serialization implemented by this adapter, which is that it doesn't use reflection, and so it should be higher performance than Gson's default serialization strategy. For large subtrees, or deployments with many discovery plugins and many subtrees, this probably results in better API performance.

Reply:
Ah. makes sense. Appreciate the explanation.